### PR TITLE
crimson/os/seastore: avoid getting wrong logical extents through "parent-invalid" lba mappings

### PR DESCRIPTION
**Pull request by xxhdx1985126**
The parent of an lba mapping may be modified after the mapping's retrieval, so we need to avoid getting extents through the mapping with an outdated view.

Signed-off-by: Xuehan Xu <xuxuehan@qianxin.com>

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [x] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

### CODE REVIEW

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test api

**Comment by xxhdx1985126** [comment]
jenkins test make check

**Comment by xxhdx1985126** [comment]
jenkins test windows

File: src/crimson/os/seastore/object_data_handler.cc
Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc:441]: But I'm still now sure whether `is_parent_valid()` is sufficient here. Seems to me if `pin` is retrived from a stable parent to a stable child, this cannot detect whether the parent has been modified, because the shared parent can still be valid. Is it correct?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/object_data_handler.cc:441]: > But I'm still now sure whether `is_parent_valid()` is sufficient here. Seems to me if `pin` is retrived from a stable parent to a stable child, this cannot detect whether the parent has been modified, because the shared parent can still be valid. Is it correct?

You're right. I think the reason that we have met this issue is that, fortunately, all mappings are retrieved from mutable lba leaf nodes when we want to `get pending logical extents` through them.

On the other hand, wrongly getting pending logical extents through stable mappings happens when those stable mappings have been updated, I think maybe we can implement the `copy_sources -> initial_pending` links mentioned in https://github.com/ceph/ceph/pull/55831#issuecomment-1990898098, so that we can check whether the stable mappings have been modified when NDEBUG is not defined, which should be able to prevent this issue. What do you think?

Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc:441]: Probably a simpler fix for now could be to pass the transaction reference to `is_parent_valid()`, because if the stable parent was modified, it must be retired in that transaction. What do you think?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/object_data_handler.cc:441]: > Probably a simpler fix for now could be to pass the transaction reference to `is_parent_valid()`, because if the stable parent was modified, it must be retired in that transaction. What do you think?

I think if the parent has been modified while the mapping hasn't been updated, that mapping should be considered valid for getting logical extents. Because, in this scenario, even if we try to get the logical extents from mutated lba leaf nodes, we still have to fall back to the stable copy sources to find the lba pointer. What do you think?

File: src/crimson/os/seastore/btree/fixed_kv_node.h
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_node.h:285]: Seems to me this is a special case due to the usage during `ObjectDataHandler::overwrite()`. So, instead of handling it in all of the get_child pathes, is it better to introduce a special interface to fix the `pos` during `ObjectDataHandler::overwrite()` when necessary ?

The dedicated interface can also be a reminder for the special case.

File: src/crimson/os/seastore/object_data_handler.cc
Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc:441]: I see, I think you are right.

IIUC the correctness of `is_parent_valid()` is under the assumption that the child (ObjectDataHandler, i.e. 
`region.pin`) is stable upon remapping. So that the lookup is still correct through the stable parent even if the parent has been invalidated (mutated or splitted) in the transaction.

If that's correct, how about adding `ceph_assert(region.pin->is_stable())` with some comments to explain ? (It's not very easy to follow because it leverages low-level properties of the lba pointer design.

Comment (cyx1231st) [reply src/crimson/os/seastore/object_data_handler.cc:441]: Also, a related unit test case is needed to prove that this keeps working when there is a related lba modification happened after getting the original pin. (Can we also validate that the original pin is wrong in this case?)

Comment (xxhdx1985126) [reply src/crimson/os/seastore/object_data_handler.cc:441]: @cyx1231st I added checks for debug builds to make sure that lba mappings haven't been modified when being traced to get logical extents. This can accurately capture the scenarios that would lead to errors, and as long as these checks are passed, there would be no problem. Please take a look, thanks.

File: src/crimson/os/seastore/btree/btree_range_pin.cc
Comment (cyx1231st) [reply src/crimson/os/seastore/btree/btree_range_pin.cc:37]: I'm still confused. If you mean the "re-get" by the newly added logic `fut = ctx.tm.get_pin()` in `do_remappings()`, but looks this happens **after** all the calls to `pin->is_stable()` or `pin->is_data_stable()`.

Probably there is another way without subtle assumptions or the low-level dependency of `!region.pin->is_parent_valid()`. Will back to it today.

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/btree_range_pin.cc:37]: > I'm still confused. If you mean the "re-get" by the newly added logic `fut = ctx.tm.get_pin()` in `do_remappings()`, but looks this happens **after** all the calls to `pin->is_stable()` or `pin->is_data_stable()`.
> 
> Probably there is another way without subtle assumptions or the low-level dependency of `!region.pin->is_parent_valid()`. I'll be back to this today.

All `pin->is_stable()` or `pin->is_data_stable()` happens before the modifications to the parent lba node at present. So in the current implementation, the assertion `assert(is_parent_valid())` holds.

Are you suggesting that we should remove this assertion from `pin->is_stable()` and `pin->is_data_stable()`?

**Comment by cyx1231st** [comment]
How about:

Introduce a public method `void PhysicalNodeMapping::attatch_extent_info()`, so that user can explicitly instruct the `pin` to attach necessary information **internally** upon getting the mapping.

So that:

Only after the "info" is attached to the `pin`, the according methods, for example:
* `Mapping::is_(data_)stable()` directly returns the attatched result.
* `TM::read_pin(_by_type)()` knows that the parent is possible to become invalid, and handle the case properly, either by:
  1. fix the pos by the parent key
  2. query the lba tree again
  3. or even smarter in the future: https://github.com/ceph/ceph/pull/55831#issuecomment-1990898098
* Which also means user can be decoupled from the details.

With `void PhysicalNodeMapping::attatch_extent_info()`, we don't need to generally assume that `pos` cannot be trusted, and the original `get_child()` doesn't need extra changes (only assert that `key == get_key_from_idx(pos)`).

@xxhdx1985126 Does it look reasonable?

File: src/crimson/os/seastore/btree/btree_range_pin.cc
Comment (cyx1231st) [reply src/crimson/os/seastore/btree/btree_range_pin.cc:37]: I'm just trying to confirm whether it is directly related to the broken issue, or it is for future proof.

> is_parent_valid() means that the parent lba node is not INVALID, an lba node modified after the pin's creation doesn't necessarily mean the lba node is INVALID.

> All pin->is_stable() or pin->is_data_stable() happens before the modifications to the parent lba node at present. So in the current implementation, the assertion assert(is_parent_valid()) holds.

^ So it's for future proof, our current code won't break without it?

Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/btree_range_pin.cc:37]: > ^ So it's for future proof, and the current code won't break without it?

Yes, that's right

**Comment by xxhdx1985126** [comment]
> How about:
> 
> Introduce a public method `void PhysicalNodeMapping::attach_extent_info()`, so that user can explicitly instruct the `pin` to attach necessary information **internally** upon getting the mapping.
> 
> So that:
> 
> Only after the "info" is attached to the `pin`, the according methods, for example:
> 
> * `Mapping::is_(data_)stable()` directly returns the attatched result.
> * `TM::read_pin(_by_type)()` knows that the parent is now possible to become invalid, and handle the case properly, either by:
>   
>   1. fix the pos by the parent key
>   2. query the lba tree again
>   3. or even smarter in the future: [crimson/os/seastore: lba pointer consolidation #55831 (comment)](https://github.com/ceph/ceph/pull/55831#issuecomment-1990898098)
> * Which also means user can be decoupled from the details.
> 
> With `void PhysicalNodeMapping::attatch_extent_info()`, we don't need to generally assume that `pos` cannot be trusted, and the original `get_child()` doesn't need extra changes (only assert that `key == get_key_from_idx(pos)`).
> 
> @xxhdx1985126 Does it look reasonable?

I'll give it try:-)

**Comment by xxhdx1985126** [comment]
@cyx1231st 

I implemented another approach in which we don't have to touch `FixedKVNode::get_child()` and `FixedKVNode::is_child_(data)_stable()` and can detect all invalid lba mapping cases.

Specifically, the main modifications are as follows:

1. assert that the parent of an lba mapping is never in the retired set of the transction
2. add `LBAMapping::parent_modified()` to check whether the parent lba leaf has been modified since the retrieval of the lba mapping
3. re-retrieve the lba mapping if the parent isn't valid anymore. Since we've made sure that the parent of an lba mapping will never be in the retried set of the transaction, this should be the only case in which we need to re-retrieve the mapping
4. fix lba mappings' `pos` if necessary before following them to get logical extents

WIth this approach, users of lba mappings, like ObjectDataHandler, OMapManager, OnodeManager, don't have to worry about the validity of the lba mappings anymore.

Please take a look, thanks.

File: src/crimson/os/seastore/btree/fixed_kv_node.h
Comment (cyx1231st) [inline src/crimson/os/seastore/btree/fixed_kv_node.h:91]: A bit of commits might be helpful to clarify the intention as the implementations might be scattered, and even be wrong accidentally:
```c++
// for the stable extent, modifications is always 0 (is that correct?)
// it will increase for each transaction-local change, so that modifications can be detected (see BtreeLBAMapping.parent_modifications)
```

**Comment by cyx1231st** [comment]
^ Still following through the commits, some initial suggestions.

File: src/crimson/os/seastore/transaction.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction.h:328]: > I'm not sure why is_retired(paddr, len) doesn't/cannot do an exact check, which can be simpler.

It can do the check, I just think searching through the retired_set for every check might be expensive in terms of cpu overhead.

File: src/crimson/os/seastore/cache.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cache.h:471]: Because stable extents with `mutation_pendings` are not in the transaction's `retired_set`

File: src/crimson/os/seastore/cache.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cache.h:467]: The extent might be pending

File: src/crimson/os/seastore/cached_extent.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:895]: Yeah, will add it.

File: src/crimson/os/seastore/transaction.h
Comment (cyx1231st) [reply src/crimson/os/seastore/transaction.h:328]: Not sure why doing an exact check is more expensive, it can still use lower_bound() but return false if paddr or len doesn't match.

This can also avoid the below issue/gap:
> If it is checking whether an arbitrary range is retired, it will be wrong if the range is retired but in two consecutive retired extents.

File: src/crimson/os/seastore/cache.h
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: Why? Looking at add_to_retired_set(ref), stable extents are always added to the `retired_set` (else case), or I'm missing anything?

File: src/crimson/os/seastore/cache.h
Comment (cyx1231st) [reply src/crimson/os/seastore/cache.h]: How about making this case explicit (instead of relying on the implicit correctness), e.g.
```c++
if (!extent->is_stable()) {
  if (extent->is_valid()) {
    // probably assert extent is local to transaction t ?
    return false;
  } else {
    return true;
  }
} else {
  ... (the stable case) ...
}
```

File: src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h
Comment (cyx1231st) [inline src/crimson/os/seastore/lba_manager/btree/btree_lba_manager.h:174]: Since there is `parent_modified()` mechanism, looks we can skip it (lookup the key-pos again) if the parent isn't modified?

File: src/crimson/os/seastore/transaction_manager.h
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction_manager.h:195]: This comment should be moved to `get_extent_if_linked()`

File: src/crimson/os/seastore/transaction.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction.h:328]: > Not sure why doing an exact check is more expensive, it can still use lower_bound() but return false if paddr or len doesn't match.
> 
> This can also avoid the below issue/gap:
> 
> > If it is checking whether an arbitrary range is retired, it will be wrong if the range is retired but in two consecutive retired extents.

Sorry, I misunderstood your question, will change it to do exact checks.

File: src/crimson/os/seastore/cache.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/cache.h]: > Why? Looking at `add_to_retired_set(ref)`, stable extents are always added to the `retired_set` (the `else` case), or I'm missing anything?

Adding an mutated extent to the `retired_set` does retire the stable one. However, mutating an extent doesn't add it to the transaction's `retired_set`.

File: src/crimson/os/seastore/transaction_manager.h
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction_manager.h:198]: move the fixed `pin` into `fut` so we don't need to distinguish them here.

File: src/test/crimson/seastore/test_transaction_manager.cc
Comment (cyx1231st) [inline src/test/crimson/seastore/test_transaction_manager.cc:2198]: Probably deserve a further test case when `pin->parent_modified()` becomes true, pin is still useful ?

**Review by cyx1231st — COMMENTED** [review]
Generally looks reasonable.

**Comment by xxhdx1985126** [comment]
@cyx1231st All comments addressed, please take a look again, thanks:-)

**Comment by cyx1231st** [comment]
@xxhdx1985126 Sorry I'm not following which comments have been addressed, the new change looks like a rebase to me, have you pushed a wrong branch?

**Comment by xxhdx1985126** [comment]
> @xxhdx1985126 Sorry I'm not following which comments have been addressed, the new change looks like a rebase to me, have you pushed a wrong branch?

Yes, sorry, I pushed the wrong branch, it's the right one now.

**Review by cyx1231st — COMMENTED** [review]
LGTM except for the below comments.

File: src/crimson/os/seastore/btree/btree_range_pin.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/btree/btree_range_pin.h:197]: yes, will do

**Comment by xxhdx1985126** [comment]
@cyx1231st All done, please take a look again, thanks.

**Comment by xxhdx1985126** [comment]
jenkins test make check

File: src/crimson/os/seastore/transaction.h
Comment (cyx1231st) [inline src/crimson/os/seastore/transaction.h:332]: Follow up https://github.com/ceph/ceph/pull/57709#discussion_r1643641271, does it look more accurate as a simple accurate check?

```c++
bool is_retired(paddr_t paddr, extent_len_t len) {
  auto iter = retired_set.lower_bound(paddr);
  if (iter == retired_set.end() ||
      (*iter)->get_paddr() != paddr) {
    return false;
  } else {
    assert(len == (*iter)->get_length());
    return true;
  }
}
```

Can you merge the 1st and the last commits for the same purpose?

**Review by cyx1231st — COMMENTED** [review]
LGTM except one last comment.

File: src/crimson/os/seastore/transaction.h
Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction.h:332]: Ah, yes, will do

Comment (xxhdx1985126) [reply src/crimson/os/seastore/transaction.h:332]: Done:-)

**Review by cyx1231st — APPROVED** [review]

**Comment by cyx1231st** [comment]
I'll merge once the build approves.

**Comment by cyx1231st** [comment]
https://shaman.ceph.com/builds/ceph/ci-yingxin-validate-pr-57709/

File: src/crimson/os/seastore/cached_extent.h
Comment (cyx1231st) [inline src/crimson/os/seastore/cached_extent.h:905]: Seems this line is failing, tracked by https://tracker.ceph.com/issues/66851

Comment (xxhdx1985126) [reply src/crimson/os/seastore/cached_extent.h:905]: > Seems this line is failing, tracked by https://tracker.ceph.com/issues/66851

This is fixed by: https://github.com/ceph/ceph/pull/58357

**Pull request merged**
